### PR TITLE
ARROW-16356: [Python] Expose RandomAccessFile::GetStream

### PR DESCRIPTION
**Pull request by alexdesiqueira**
This PR exposes `RandomAccessFile::GetStream` to `pyarrow` as `get_stream`, as requested in [JIRA#ARROW-16356](https://issues.apache.org/jira/browse/ARROW-16356).

### CODE REVIEW

**Comment by alexdesiqueira** [comment]
@pitrou this is what I could do so far. Please let me know if I'm on the right track — or how far I am from it :) Thanks!

File: python/pyarrow/includes/libarrow.pxd [hunk 1]
```diff
@@ -1262,6 +1262,11 @@ cdef extern from "arrow/io/api.h" namespace "arrow::io" nogil:
                                                                   Seekable):
         CResult[int64_t] GetSize()
 
+        CResult[shared_ptr[CInputStream]] GetStream(
```
Comment (pitrou) [inline python/pyarrow/includes/libarrow.pxd:1265]: Note that `GetStream` is a static C++ method, so you must reflect that in the Cython declaration.

File: python/pyarrow/io.pxi [hunk 2]
```diff
@@ -395,6 +395,38 @@ cdef class NativeFile(_Weakrefable):
...
+            int64_t c_file_offset
+            int64_t c_nbytes
+
+        c_file_offset = file_offset
+        c_nbytes = nbytes
+
+        handle = self.get_random_access_file()
+
+        data = <shared_ptr[CInputStream]> GetResultValue(
+            handle.get().GetStream(self.read(), c_file_offset, c_nbytes))
```
Comment (pitrou) [inline python/pyarrow/io.pxi:426]: `GetStream` being a static method, you shouldn't call it on an instance. Also, I'm curious why you're passing `self.read()` here? The goal is _not_ to read data immediately.

File: python/pyarrow/tests/test_io.py [hunk 3]
```diff
@@ -125,6 +125,22 @@ def test_python_file_read():
...
+    data = b'data1data2data3data4data5'
+
+    buf = BytesIO(data)
+    f = pa.PythonFile(buf, mode='r')
+
+    stream1 = f.get_stream(file_offset=0, nbytes=10)
+    stream2 = f.get_stream(file_offset=9, nbytes=16)
+
+    buf_nbytes6 = stream1.read(nbytes=6)
+    assert buf_nbytes6 == b'data1d'
```
Comment (pitrou) [inline python/pyarrow/tests/test_io.py:138]: Probably want to exercise this a bit more. When it the end of file reached?

**Comment by pitrou** [comment]
@alexdesiqueira Did you manage to get a working PyArrow development setup? This would be easier than waiting for CI to run :-)

**Comment by alexdesiqueira** [comment]
> @alexdesiqueira Did you manage to get a working PyArrow development setup? This would be easier than waiting for CI to run :-)

@pitrou thanks for reviewing! You're right :) I was setting it up but had an issue with `ASan` that I couldn't solve yet:
```sh
==181460==ASan runtime does not come first in initial library list;
you should either link runtime to your application or manually preload it with LD_PRELOAD.
```

I'll check `ci/docker` and `python/examples/minimal_build` to setup a container for tests today.
Thanks again!

File: python/pyarrow/tests/test_io.py (hunk 3 above)
Comment (alexdesiqueira) [reply python/pyarrow/tests/test_io.py:138]: Got it. I'll get back to that. Thanks!

**Comment by alexdesiqueira** [comment]
@pitrou there are some things I don't understand yet, and I'll check them later to continue working on this. Thanks!

**Comment by pitrou** [comment]
@alexdesiqueira Did you manage to get a working local build? It seems like you are waiting for CI to test your changes.

**Comment by alexdesiqueira** [comment]
@pitrou I did this night :sweat_smile: I just pushed some changes I had here, sorry about that.

File: python/pyarrow/io.pxi [hunk 4]
```diff
@@ -395,6 +395,42 @@ cdef class NativeFile(_Weakrefable):
 
         return PyObject_to_object(obj)
 
+    def get_stream(self, file_offset, nbytes):
+        """
+        Returns an input stream that reads a file segment independent of the
+        state of the file.
```
Comment (pitrou) [inline python/pyarrow/io.pxi:401]: Nit:
```suggestion
        Return an input stream that reads a file segment independent of the
        state of the file.
```

File: python/pyarrow/io.pxi [hunk 5]
```diff
@@ -395,6 +395,42 @@ cdef class NativeFile(_Weakrefable):
...
+        without interfering with each other.
+
+        Parameters
+        ----------
+        file_offset : int
+        nbytes : int
+
+        Returns
+        -------
+        stream : bytes
```
Comment (pitrou) [inline python/pyarrow/io.pxi:413]: The annotation here is incorrect, should be something like:
```suggestion
        stream : NativeFile
```

File: python/pyarrow/io.pxi [hunk 6]
```diff
@@ -395,6 +395,42 @@ cdef class NativeFile(_Weakrefable):
 
         return PyObject_to_object(obj)
 
+    def get_stream(self, file_offset, nbytes):
```
Comment (pitrou) [inline python/pyarrow/io.pxi:398]: This implementation seems basically correct to me, thanks.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update and sorry for the delay @alexdesiqueira . The implementation is looking good, some further comments below.

**Comment by pitrou** [comment]
@alexdesiqueira Are you willing to update this PR?

**Comment by alexdesiqueira** [comment]
@pitrou yes, but I won't have too much time this week. Let me know if that's okay.

**Comment by pitrou** [comment]
Yes, it's okay :-)

**Comment by alexdesiqueira** [comment]
I'm back @pitrou :slightly_smiling_face: 
For these tests, I'm trying to emulate [cpp/src/arrow/io/memory_test.cc:320](https://github.com/apache/arrow/blob/8daa7a4ed5629c0020dadf7325a6b523bdfc62e9/cpp/src/arrow/io/memory_test.cc#L320):

```c++
// Read to end of each stream
  ASSERT_OK_AND_EQ(2, stream1->Read(4, buf3));
  ASSERT_EQ(0, std::memcmp(buf3, "a2", 2));
  ASSERT_OK_AND_EQ(10, stream1->Tell());

  ASSERT_OK_AND_EQ(0, stream1->Read(1, buf3));
  ASSERT_OK_AND_EQ(10, stream1->Tell());

  // stream2 had its extent limited
  ASSERT_OK_AND_ASSIGN(buf2, stream2->Read(20));
  ASSERT_TRUE(SliceBuffer(buf, 13, 12)->Equals(*buf2));

  ASSERT_OK_AND_ASSIGN(buf2, stream2->Read(1));
  ASSERT_EQ(0, buf2->size());
  ASSERT_OK_AND_EQ(16, stream2->Tell());

  ASSERT_OK(stream1->Close());

  // idempotent
  ASSERT_OK(stream1->Close());
  ASSERT_TRUE(stream1->closed());

  // Check whether closed
  ASSERT_RAISES(IOError, stream1->Tell());
  ASSERT_RAISES(IOError, stream1->Read(1));
  ASSERT_RAISES(IOError, stream1->Read(1, buf3));
```
What do you think on what we have so far? I'm kinda lost on the ones we have missing. Any ideas I could use?
Thanks!

File: python/pyarrow/tests/test_io.py [hunk 7]
```diff
@@ -125,6 +125,40 @@ def test_python_file_read():
...
+
+    buf_stream1_6 = stream1.read(nbytes=6)
+    assert len(buf_stream1_6) == 6
+    assert buf_stream1_6 == b'data1d'
+    assert stream1.tell() == 6
+
+    # Read to end of each stream
+    buf_stream1_4 = stream1.read(nbytes=4)
+    assert len(buf_stream1_4) == 2
+    assert buf_stream1_4 == b'a2'
```
Comment (pitrou) [inline python/pyarrow/tests/test_io.py:150]: `buf_stream1_4` should be `b'ata2'`, no?

**Comment by alexdesiqueira** [comment]
@pitrou sorry about the delay also; I'll try to fix it now.

**Comment by alexdesiqueira** [comment]
@pitrou sorry about the delay as well, life got in the way :slightly_smiling_face:

**Comment by pitrou** [comment]
It's fine @alexdesiqueira, but there are CI failures, are you willing to tackle these (I'm assuming you should be able to reproduce them locally)?

**Comment by alexdesiqueira** [comment]
Hey @pitrou,
I'm sorry, life got in the way and I'm postponing/dropping a lot of extra work. Could I ask you to take over? I'll assume it would be way easier for you to finish this :slightly_smiling_face: 
Apologies again.

**Comment by alexdesiqueira** [comment]
@pitrou thank you! I appreciate it. Thanks for the reviews also; I learned a lot in the process.

**Pull request closed without merging**